test/test_torch.py

```diff
@@ -16330,16 +16329,140 @@ def test_svd_tall_some_col_maj(self, device, dtype):
+    'sin': ufunc_meta(is_floating=True),
+    'sinh': ufunc_meta(is_floating=True),
+    'sqrt': ufunc_meta(is_floating=True),
+    'trunc': ufunc_meta(is_floating=True),
+    'digamma': ufunc_meta(is_floating=True, complex_on=()),
+    'rsqrt': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'sigmoid': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'atan': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'cos': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'cosh': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'exp': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'tan': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'tanh': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'erf': ufunc_meta(is_floating=True, complex_on=()),
+    'erfc': ufunc_meta(is_floating=True, complex_on=()),
+    'erfinv': ufunc_meta(is_floating=True, complex_on=()),
+    'lgamma': ufunc_meta(is_floating=True, complex_on=())
+}
+
+# Creates and adds a test of the given unary floating ufunc's type promotion
+# Unary floating ufuncs promote tensors of bool and integral types to
+# the default scalar type.
+def generate_unary_floating_ufunc_promo_test(cls, op_str):
+    int_types = [torch.bool, torch.uint8, torch.int8, torch.int16,
+                 torch.int, torch.long]
+    float_types = [torch.float, torch.double]
+    float_types_cuda = [torch.half, torch.float, torch.double]
+    complex_types = [torch.complex64, torch.complex128]
+    default_types = [torch.float, torch.double]
+    default_types_cuda = [torch.half, torch.float, torch.double]
+
+    def test_fn(self, device):
+        op = getattr(torch, op_str)
+
+        my_default_types = default_types
+        my_float_types = float_types
+        if self.device_type == 'cuda':
+            my_default_types = default_types_cuda
+            my_float_types = float_types_cuda
+        if self.device_type in unary_ufuncs[op_str].complex_on:
+            my_float_types.extend(complex_types)
+
+        prev_def_type = torch.get_default_dtype()
+        try:
+            for in_type, def_type in product(int_types, my_default_types):
+                torch.set_default_dtype(def_type)
+                t = torch.tensor((1,), device=device, dtype=in_type)
+                self.assertEqual(op(t).dtype, def_type)
+        finally:
+            torch.set_default_dtype(prev_def_type)
+
+        for in_type in my_float_types:
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            self.assertEqual(op(t).dtype, in_type)
+
+        only_cpu_ops = ['atan', 'cos', 'cosh', 'erf', 'erfc', 'exp', 'tan', 'tanh']
+        for in_type in my_float_types + int_types:
+            if(op_str in only_cpu_ops and self.device_type == 'cuda'):
+                continue
+            t = torch.tensor((1,), device=device, dtype=in_type)
+            for out_type in my_float_types:
+                if(in_type == torch.bool and out_type in complex_types):
```

Comment:
What happens when PyTorch tries to promote bool to complex?

Reply:
I think this is because of the common dtype calculation in the TensorIterator logic. `torch.cos(BoolTensor, out=<tensor with complex64>)` throws `RuntimeError: dtype != ScalarType::Undefined INTERNAL ASSERT FAILED at /home/krshrimali/Documents/pytorch/pytorch/aten/src/ATen/native/TensorIterator.cpp:124, please report a bug to PyTorch. `.

Reply:
I thought to dig further on this error.

For the common dtype calculation, the function `promote_skip_undefined` (at: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TypeProperties.cpp#L64) uses promotion table in PyTorch (https://github.com/pytorch/pytorch/blob/master/c10/core/ScalarType.h#L438). 

Here, for bool (`b1`) to complex types (`c2`, `c4`, `c8`), the output type is `ud` (undefined). And then we have an assert (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorIterator.cpp#L124) which throws this error.